feat(router): new output that would notify when link is activated

anandtiwary opened the pull request:
This commit adds a new output to `routerLinkActive` directive.
Whenever, the associated link becomes active or inactive, an
event will be fired on this out with the correct status

PR Close #37284

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #37284


## What is the new behavior?

Introduced a new Output that emits when link gets activated or inactivated

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

atscott committed:
fix(router): null/undefined routerLink should disable navigation (#43087)

The current behavior of `routerLink` for `null` and `undefined` inputs is to treat
the input the same as `[]`. This creates several unresolvable issues with
correctly disabling the links because `commands = []` does _not_ behave the same
as disabling a link. Instead, it navigates to the current page, but will also
clear any fragment and/or query params.

The new behavior of the `routerLink` input will be to completely disable navigation
for `null` and `undefined` inputs. For HTML Anchor elements, this will also mean
removing the `href` attribute.

Fixes #21457
Fixes #13980
Fixes #31154

BREAKING CHANGE:
Previously `null` and `undefined` inputs for `routerLink` were
equaivalent to empty string and there was no way to disable the link's
navigation.
In addition, the `href` is changed from a property `HostBinding()` to an
attribute binding (`HostBinding('attr.href')`). The effect of this
change is that `DebugElement.properties['href']` will now return the
`href` value returned by the native element which will be the full URL
rather than the internal value of the `RouterLink` `href` property.

PR Close #43087

danieltre23 committed:
refactor(compiler-cli): export `ExtendedTemplateCheckerImpl` and template checks (#43107)

This commit exports the implementation of `ExtendedTemplateChecker` to
generate extended template diagnostics and all the template checks.

Refs #42966

PR Close #43107

dario-piotrowicz committed:
refactor(docs-infra): use tslint in aio's tools-lint script (#43010)

Instead of the deprecated tslint use eslint in the aio's tools-lint
script

PR Close #43010

danieltre23 committed:
refactor(compiler-cli): add `extendedTemplateCheck` phase to compiler (#43107)

This commit integrates extended template checks with the compiler, by
adding another phase of diagnostics generation. This integration is
under the `_extendedTemplateDiagnostics` flag.

Refs #42966

PR Close #43107

josmar-crwdstffng committed:
docs: update links and formatting (#43066)

Update links and formatting prior to creating sub-topics.

Update links to reference links for consistency.  Update formatting for code sections.  Updated list alignment for consistent formatting.

PR Close #43066

danieltre23 committed:
test(compiler-cli): add extended template diagnostics end-to-end tests (#43107)

This commit adds extended template diagnostics end-to-end tests, to make
sure the diagnostics are generated correctly. Template checks are
already tested with unit tests.

Refs #42966

PR Close #43107

danieltre23 committed:
test(language-service): add extended diagnostics test (#43107)

Add a test in the langauge-service to make sure the extended template
diagnostics are being correctly generated.

Refs #42966

PR Close #43107

devversion committed:
feat(compiler-cli): inline resources when generating class metadata calls (#43178)

Previously with View Engine output, the `enableResourceInlining` option
could be set to inline external templates and styles (also for the
resulting `.metadata.json` files). We want to do the same for the Ivy
compilation pipeline (regardless of the compilation mode). The full
compilation definitions, and partial declarations currently already
inline resources in a way that no external requests need to be made.

Although there is one exception currently. These are the calls for
setting class metadata (for testbed overrides). This commit updates
the set class metadata calls (for both partial and full compilation)
to always inline resources. This means that libraries do not need
to start shipping external styles/templates just for the
`setClassMetadata` calls.

Note: Only doing this for partial compilation has been considered, but
it seems like it would be simpler implementation-wise to do this for
full compilation as well. Given the external resources are already
inlined (through their `ecmp` definitions), it seems acceptable (or
even more aligned) to do the same for the set class metadata calls.

PR Close #43178

atscott committed:
feat(core): Add migration to update empty routerLinks in templates (#43176)

The previous behavior of `RouterLink` for `null` and `undefined` inputs was to treat
the input the same as `[]` or `''`. This creates several unresolvable issues with
correctly disabling the links because `commands = []` does not behave the same
as disabling a link. Instead, it navigates to the current page, but will also
clear any fragment and/or query params.

The new behavior of the `routerLink` input will be to completely disable navigation
for `null` and `undefined` inputs. For HTML Anchor elements, this will also mean
removing the `href` attribute.

Migration for change in #43087

PR Close #43176

renovate-bot committed:
build: update bazel to v4.0.0-rc.0 (#43138)

PR Close #43138

devversion committed:
build: update rules_nodejs version for bazel integration test (#43138)

To keep our integration tests aligned with the setup we use
for the Angular repo and Components repo, we should also update
the `rules_nodejs` version for our `bazel` integration test.

PR Close #43138

danieltre23 committed:
refactor(compiler-cli): add `ExtendedTemplateChecker` (#43107)

Change the current way to run template checks to the
`ExtendedTemplateChecker` instead of just the
`getExtendedTemplateDiagnosticsForComponent` function. Refactored the
tests that used the previous function to use the new class.

Refs #42966

PR Close #43107

anandtiwary committed:
feat(router): new output that would notify when link is activated

This commit adds a new output to `routerLinkActive` directive.
Whenever, the associated link becomes active or inactive, an
event will be fired on this out with the correct status

PR Close #37284

anandtiwary committed:
feat(router): new output that would notify when link is activated

This commit adds a new output to `routerLinkActive` directive.
Whenever, the associated link becomes active or inactive, an
event will be fired on this out with the correct status

PR Close #37284

serpent5 committed:
docs: tidy up description of NgComponentOutlet (#43159)

PR Close #43159

atscott commented on packages/router/src/directives/router_link_active.ts:
```diff
@@ -114,35 +124,42 @@ export class RouterLinkActive implements OnChanges, OnDestroy, AfterContentInit
-  set routerLinkActive(data: string[]|string) {
-    const classes = Array.isArray(data) ? data : data.split(' ');
-    this.classes = classes.filter(c => !!c);
+  set routerLinkActive(data: string[]|string|undefined) {
```
Why was `undefined` added here? It doesn't seem related to this change.

anandtiwary replied:
Let me remove it

atscott commented on packages/router/src/directives/router_link_active.ts:
```diff
@@ -99,6 +108,7 @@ export class RouterLinkActive implements OnChanges, OnDestroy, AfterContentInit
    */
   @Input() routerLinkActiveOptions: {exact: boolean}|IsActiveMatchOptions = {exact: false};
 
+  @Output() public readonly routerLinkActivated: EventEmitter<boolean> = new EventEmitter();
```
```suggestion
  @Output() readonly routerLinkActivated: EventEmitter<boolean> = new EventEmitter();
```
nit: `public` is not necessary

atscott replied:
```suggestion
  @Output() public readonly isActiveChange: EventEmitter<boolean> = new EventEmitter();
```
Bike-shedding: I'm not sure how much I like "routerLinkActivated". How about `isActiveChange`?

atscott commented on packages/router/src/directives/router_link_active.ts:
```diff
@@ -163,24 +180,28 @@ export class RouterLinkActive implements OnChanges, OnDestroy, AfterContentInit
           }
         });
+
+        // Emit on routerLinkActivated after classes are updated
+        this.routerLinkActivated.emit(hasActiveLinks);
```
This is going to emit more frequently than we would like. It should only emit when the value changes since the last emitted value.

anandtiwary replied:
It is part of `if` at L173. So, won't it only emit when the two values are not matching?

atscott replied:
Ah, you're right :)

atscott reviewed (changes requested):
I mentioned this in the previous PR - could you separate the unrelated styling changes into a different commit or PR and limit your changes to only what's necessary for this feature?

anandtiwary commented:
@atscott Some styling came from the git hook prettier. I can take care of other comments but I don't think i can revert all the formatting related changes. Let me know if you think otherwise

atscott commented:
> @atscott Some styling came from the git hook prettier. 

Is this a git hook local to your machine? We use clang-format, not prettier, to format TS files.

anandtiwary commented:
interesting, I do have a git hook and prettier plugin in vscode. 

If angular doesn't use prettier, how come we have `https://github.com/angular/angular/blob/master/.prettierrc` and prettier in the dev dependencies?

atscott commented:
> I do see that angular has `https://github.com/angular/angular/blob/master/.prettierrc` and prettier in the dev dependencies? Are they used to format some other type of files?

That's right. We use prettier to format yaml files.

anandtiwary commented:
Got it.. Thanks for your reply. This is my first PR to angular, so i am facing some tooling related issues. Let me fix those and upload a new commit.

erik-slack committed:
docs: fix typo (#43190)

Fix typo in description of ngxp and link for angular community meetup

PR Close #43190

dylhunn committed:
docs: release notes for the v12.2.2 release (#43196)

josephperrott committed:
release: cut the v13.0.0-next.2 release (#43200)

anandtiwary committed:
refactor(router): addressing review comments

- renamed output property
- removed public keyword
- updated tests
- reverted formatting changes

anandtiwary committed:
refactor(router): addressing review comments

- renamed output property
- removed public keyword
- updated tests
- reverted formatting changes

anandtiwary committed:
refactor(router): resolving merge conflicts

- resolcing merge conflicts

anandtiwary commented:
hmm.. So i got a build failure that branch rebase failed. So, I pulled latest master in my forked repo's master and then rebased my local `router-activated-37284` with master. And then i pushed. The logs were looking fine when i checked locally.  What did i do wrong?

atscott commented:
@anandtiwary I do not know what you did with your rebase. It's probably easiest to start over again.
